cosima_cookbook/explore.py

```diff
@@ -0,0 +1,965 @@
+        )
+
+        # Date selection widget
+        self.widgets['daterange'] = widgets.SelectionRangeSlider(
+            options=['0000','0001'],
+            index=(0,1),
+            description='Date range',
+            layout={'width': '80%'},
+            disabled=True
+        )
+
+        # Variable filter selector combo widget. Pass in two widgets so they
+        # can be updated by the VariableSelectorInfo widget
+        self.widgets['var_selector'] = VariableSelectorInfo(self.de.variables, 
+                                                            daterange=self.widgets['daterange'],
+                                                            frequency=self.widgets['frequency'],
+                                                            rows=20)
+
+        # DataArray information widget
+        self.widgets['data_box'] = widgets.HTML()
+
+        # Data load button
+        self.widgets['load_button'] = Button(
+            description='Load',
+            disabled=False,
+            layout={'width': '20%', 'align': 'center'},
+            tooltip='Click to load data'
+        )
+
+        info_pane = VBox([self.widgets['frequency'],
+                          self.widgets['daterange']],
+                          layout={'padding': '10% 0', 'width': '50%'})
+
+        centre_pane = HBox([VBox([self.widgets['var_selector']]),
+                                  info_pane])
+
+        # Call super init and pass widgets as children
+        super().__init__(children=[self.widgets['header'],
+                                   self.widgets['expt_selector'],
+                                   centre_pane,
+                                   self.widgets['load_button'],
+                                   self.widgets['data_box']])
+
+    def _set_handlers(self):
+        """
+        Define routines to handle button clicks and experiment selection
+        """
+
+        self.widgets['load_button'].on_click(self._load_data)
+        self.widgets['expt_selector'].observe(self._expt_eventhandler, names='value')
+
+    def _expt_eventhandler(self, selector):
+        """
+        Called when experiment dropdown menu changes
+        """
+        self._load_experiment(selector.new)
+
+    def _load_data(self, b):
+        """
+        Called when load_button clicked
+        """
+
+        data_box = self.widgets['data_box']
+
+        varname = self.widgets['var_selector'].get_selected()
+        (start_time, end_time) = self.widgets['daterange'].value
+        frequency = self.widgets['frequency'].value
+
+        load_command = """
+        <pre><code>cc.querying.getvar('{expt}', '{var}', session, 
+                    start_time='{start}', end_time='{end}', frequency='{frequency}')</code></pre>
+        """.format(expt=self.widgets['expt_selector'].value, 
+                var=varname,
+                start=str(start_time),
+                end=str(end_time),
+                frequency=str(frequency))
```

Comment:
(mostly curious) do you need to explicitly `str()` arguments to `format`? Do you get their `repr` instead?

Reply:
Yes that did seem to be necessary. I recall it failed when I used native datetime objects when I removed the explicit conversion. I assumed there was some sqlalchemy jiggery pokery going on and bunged them back on.